### ISSUE TITLE
이미지 분류

### ISSUE THREAD

**Issue by rkdogo08** [rkdogo08/Quiz#1]
## 학습 데이터 준비
*  함수`resize32()`은 opencv를 이용하여 경로`dir`에 저장된 이미지를 불러와 1D Vector로 변환하는 함수이다. `cv2.imread(dir)`를 이용하여 경로 `dir`에 저장된 이미지를 열고 RGB 3차원으로 저장된 이미지를 흑백 1차원으로 변환한다. `resize()`를 이용하여 1024x1인 1D Vector로 변환한다.
``` 
def resize32(dir):
    img= cv2.imread(dir)
    img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    img_gray.resize(32*32)
    return img_gray
```

* os.listdir()를 이용하여 이미지가 저장되어 있는 폴더의 하위 디렉토리를 검색하여 이미지 파일의 이름을 저장한다.
plt.subplots()를 이용하여 데이터의 이미지를 시각화하고 데이터의 라벨을 그림의 라벨로 지정한다.
파일명은 dog.560.jpg와 같이 저장되어 있으므로 split('.')를 이용하여 ‘.’을 기준으로 분리하여 라벨을 얻어낸다. 얻어낸 라벨이 ‘cat’이면 0, ‘dog’이면 1으로 라벨링한다.
이미지 데이터는 함수`resize32()`를 이용하여 1D Vector로 변환한다.

```
dataset_train = "../input/2019-fall-pr-project/train/train/"

Y=[]
x=[]
data=os.listdir(dataset_train)
fig,ax=plt.subplots(3,3)
for i,axi in enumerate(ax.flat):
    img= cv2.imread(dataset_train+data[i])
    axi.imshow(cv2.cvtColor(img, cv2.COLOR_BGR2RGB))
    axi.set(xticks=[], yticks=[],xlabel=data[i].split('.')[0])

for i in range(10000):
    Y.append(0 if data[i].split('.')[0]=='cat' else 1)
    image=resize32(dataset_train+data[i])
    x.append(image)

X=np.array(x)
```
* Train set과 validation set은 0.75 : 0.25로 나누어 학습을 진행하였다.

```
X_train,X_val,y_train,y_val=train_test_split(X,Y,test_size=0.25,random_state=42)
```

## KNeighborsClassifier과 GridSearchCV를 이용한 학습기반 모델 설계

* KNeighborsClassifier를 이용하여 모델을 설계하였다. `n_neighbors`는 결과 예측시 몇개의 이웃 데이터를 가지고 투표할 것인가에 대한 파라미터이다. GridSearchCV를 이용하여 가장 정확도가 높은 `n_neighbors` 파라미터를 구하고 `best_estimator_`를 통해 모델을 저장한다.
```
from sklearn.model_selection import GridSearchCV

knn = KNeighborsClassifier()
param_grid = {'n_neighbors': np.arange(1, 10)}
grid = GridSearchCV(knn, param_grid, cv=5)
grid.fit(X_train,y_train)
model=grid.best_estimator_
yfit=model.predict(X_val)
```

## 테스트 데이터 가공
* 테스트 데이터가 저장된 폴더의 하위 파일을 `dataset_test`에 저장하고 이를 함수`resize32()`를 이용해 1D Vector로 변환해준다.
```
dataset_test = "../input/2019-fall-pr-project/test1/test1/"

x=[]
files=os.listdir(dataset_test)
for i in range(5000):
    image=resize32(dataset_test+files[i])
    x.append(image)
test_x=np.array(x)
```
## 학습한 모델을 이용하여 결과 예측
* 학습된 모델을 이용하여 테스트 데이터를 라벨링한다.
```
result=model.predict(test_x)
```

## 결과 파일 저장
* 캐글 제출 형식을 맞추기 위해 0번째 행의 index 이름을 `"id"`로 바꾸고 데이터는 'label'`로 바꿔준다. 마지막으로 data frame을 csv파일로 변환하여 저장한다.
```
import pandas as pd
result=np.append(['label'],result)
df = pd.DataFrame(result,columns=[' '])
df=df.rename(index={0:"id"})

df.to_csv('results.csv',index=True, header=False)
```
